### PR TITLE
Refactor and bugfix AMD turbulence model

### PR DESCRIPTION
**Pull request by moprak-nrel**
This PR has a few different things:

1. The AMD model is now refactored to use `fvm::gradient` instead of hand computing it, this allows the model calculations to also look just like the tensor expression
2. There was a bug in the buoyancy correction term in the turbulent viscosity calculation, this is now fixed
3. Added a member function to expose the full line derivative vector in `FieldPlaneAveraging`, currently the only access to the derivative is through single point functions. 
4. The unit test has been updated to incorporate the bugfix

TODO: 
- [x] Update documentation
- [x] Test with GPU

### CODE REVIEW

File: amr-wind/turbulence/LES/AMD.H [hunk 1]
```diff
@@ -45,179 +53,85 @@ private:
...
-    amrex::Real num_buoy =
-        C * beta * (dx2 * wx * tx + dy2 * wy * ty + dz2 * wz * tz);
-    amrex::Real denom = (ux * ux + uy * uy + uz * uz + vx * vx + vy * vy +
-                         vz * vz + wx * wx + wy * wy + wz * wz) +
-                        1e-15;
+    amrex::Real num_shear = 0;
+    amrex::Real num_buoy = 0;
+    amrex::Real denom = 0;
+    const int n_ind = idxOp(i, j, k);
+    amrex::Real h = nlo + (n_ind + 0.5) * dx[normal_dir];
```
Comment (marchdf) [inline amr-wind/turbulence/LES/AMD.H:86]: I think there's a bunch you can `const` here: `h`, `gradTbar_h`, `diuj`, `sij`, `dkui` etc ?

File: amr-wind/turbulence/LES/AMD.H [hunk 2]
```diff
@@ -45,179 +53,85 @@ private:
...
+        amr_wind::interp::linear(gradTbar_coord, gradTbar, h);
+    for (int ii = 0; ii < AMREX_SPACEDIM; ++ii) {
+        // num_buoy = dw/dx_i dT/dx_i
+        num_buoy =
+            num_buoy + gradVel(i, j, k, normal_dir * AMREX_SPACEDIM + ii) *
+                           (gradT(i, j, k, ii) - gradTbar_h) * dx[ii] * dx[ii];
+        for (int jj = 0; jj < AMREX_SPACEDIM; ++jj) {
+            amrex::Real diuj = gradVel(i, j, k, ii * AMREX_SPACEDIM + jj);
+            amrex::Real djui = gradVel(i, j, k, jj * AMREX_SPACEDIM + ii);
+            denom = denom + diuj * diuj;
```
Comment (marchdf) [inline amr-wind/turbulence/LES/AMD.H:97]: use `+=` ?

File: amr-wind/turbulence/LES/AMD.H [hunk 3]
```diff
@@ -45,179 +53,85 @@ private:
...
+            amrex::Real sij = 0.5 * (diuj + djui);
+            for (int kk = 0; kk < AMREX_SPACEDIM; ++kk) {
+                amrex::Real dkui = gradVel(i, j, k, ii * AMREX_SPACEDIM + kk);
+                amrex::Real dkuj = gradVel(i, j, k, jj * AMREX_SPACEDIM + kk);
+                num_shear = num_shear + dkui * dkuj * dx[kk] * dx[kk] * sij;
+            }
+        }
+    }
+    denom = std::max(1e-15, denom);
+    num_shear = -C * num_shear;
```
Comment (marchdf) [inline amr-wind/turbulence/LES/AMD.H:107]: for these, use `*=` ?

Comment (marchdf) [reply amr-wind/turbulence/LES/AMD.H:107]: These comments go for the `amd_thermal_diff` bellow too.

File: amr-wind/turbulence/LES/AMD.cpp [hunk 4]
```diff
@@ -42,6 +46,27 @@ void AMD<Transport>::parse_model_coeffs()
 template <typename Transport>
 void AMD<Transport>::update_turbulent_viscosity(
     const FieldState fstate, const DiffusionType /*unused*/)
+{
+    switch (m_normal_dir) {
```
Comment (marchdf) [inline amr-wind/turbulence/LES/AMD.cpp:50]: How about: 
```
cont auto dir_op = (m_normal_dir == 0) ? XDir() : ( (m_normal_dir == 1)? YDir() : ZDir());
update_turbulent_viscosity(fstate,  dir_op);
```

File: amr-wind/turbulence/LES/AMD.cpp [hunk 5]
```diff
@@ -53,168 +78,54 @@ void AMD<Transport>::update_turbulent_viscosity(
...
-        const amrex::Real dx = geom.CellSize()[0];
-        const amrex::Real dy = geom.CellSize()[1];
-        const amrex::Real dz = geom.CellSize()[2];
-
+        const auto& problo = geom.ProbLoArray();
+        const amrex::Real nlo = problo[normal_dir];
+        const auto& dx = geom.CellSizeArray();
         for (amrex::MFIter mfi(mu_turb(lev)); mfi.isValid(); ++mfi) {
             const auto& bx = mfi.tilebox();
+            const auto& gradVel_arr = (*gradVel)(lev).array(mfi);
```
Comment (marchdf) [inline amr-wind/turbulence/LES/AMD.cpp:116]: some of these can be `const_array`

File: amr-wind/turbulence/LES/AMD.cpp [hunk 6]
```diff
@@ -53,168 +78,54 @@ void AMD<Transport>::update_turbulent_viscosity(
...
-        const auto& domain = geom.Domain();
-
-        const amrex::Real dx = geom.CellSize()[0];
-        const amrex::Real dy = geom.CellSize()[1];
-        const amrex::Real dz = geom.CellSize()[2];
-
+        const auto& problo = geom.ProbLoArray();
+        const amrex::Real nlo = problo[normal_dir];
+        const auto& dx = geom.CellSizeArray();
         for (amrex::MFIter mfi(mu_turb(lev)); mfi.isValid(); ++mfi) {
```
Comment (marchdf) [inline amr-wind/turbulence/LES/AMD.cpp:114]: wanna learn something new and do the fused parallelfor? It's more fun.

**Review by marchdf — CHANGES_REQUESTED** [review]

File: amr-wind/turbulence/LES/AMD.H [hunk 7]
```diff
@@ -45,179 +53,84 @@ private:
...
     int i,
     int j,
     int k,
-    amrex::Real dx, // Grid spacing in x direction
-    amrex::Real dy, // Grid spacing in x direction
-    amrex::Real dz, // Grid spacing in x direction
-    amrex::Real C,  // Poincare const
-    amrex::Array4<amrex::Real const> const& vel,
-    amrex::Array4<amrex::Real const> const& theta) noexcept
+    const amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> dx, // Grid spacing
```
Comment (marchdf) [inline amr-wind/turbulence/LES/AMD.H:115]: you can pass by ref `&` for gpuarray

File: amr-wind/turbulence/LES/AMD.cpp [hunk 8]
```diff
@@ -53,168 +57,54 @@ void AMD<Transport>::update_turbulent_viscosity(
     const auto& den = m_rho.state(fstate);
     const auto& geom_vec = repo.mesh().Geom();
     amrex::Real beta = 0.0;
+    const int normal_dir = m_normal_dir;
```
Comment (marchdf) [inline amr-wind/turbulence/LES/AMD.cpp:60]: are you using `normal_dir` on GPU? If not, then this isn't necessary

Comment (marchdf) [reply amr-wind/turbulence/LES/AMD.cpp:60]: oh you are, nvm

File: amr-wind/utilities/DirectionSelector.H [hunk 9]
```diff
@@ -108,4 +108,13 @@ ParallelBox(const amrex::Box& bx, const amrex::IntVect& iv)
     return lbx;
 }
 
+namespace amr_wind {
+
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE int
+direction_selector(int i, int j, int k, int dir)
```
Comment (marchdf) [inline amr-wind/utilities/DirectionSelector.H:114]: const all these?

**Comment by moprak-nrel** [comment]
Comparing the old/fixed AMD behavior against Smagorisnky: [amd_line_plots.pdf](https://github.com/Exawind/amr-wind/files/14983962/amd_line_plots.pdf). There is still a significant contribution from the buoyancy term in `\nu_t` near sharp `\theta` changes, but it's less pathological than the previous behavior.

File: docs/sphinx/theory/theory.rst [hunk 10]
```diff
@@ -287,97 +287,6 @@ The anisotropic derivative is used to define the eddy viscosity as
        D_e &= \frac{- (\hat{\partial}_k \widetilde{u}_i) (\hat{\partial}_k \widetilde{\Theta}) \partial_i \widetilde{\Theta} }{(\partial_l \widetilde{\Theta}) (\partial_l \widetilde{\Theta})}
    \end{aligned}
 
-- **Implementation details:**
```
Comment (marchdf) [inline docs/sphinx/theory/theory.rst:290]: This is all removed intentionally?

File: amr-wind/utilities/linear_interpolation.H [hunk 11]
```diff
@@ -79,31 +80,27 @@ find_index(const It begin, const It end, const T& x, const int hint = 1)
...
-    }
-
-    amrex::Abort("idx undefined");
-    exit(1);
+    static constexpr DType1 eps = 1.0e-8;
+    int j = idx.idx;
+    const auto denom = (xbegin[j + 1] - xbegin[j]);
+    const auto facR = (denom > eps) ? ((xout - xbegin[j]) / denom) : 1.0;
+    const auto facL = static_cast<DType1>(1.0) - facR;
+    const auto yout = facL * yinp[j] + facR * yinp[j + 1];
```
Comment (marchdf) [inline amr-wind/utilities/linear_interpolation.H:102]: can just return this without storing into `yout`?

File: amr-wind/utilities/linear_interpolation.H [hunk 12]
```diff
@@ -79,31 +80,27 @@ find_index(const It begin, const It end, const T& x, const int hint = 1)
...
-        const auto facR = (denom > eps) ? ((xout - xbegin[j]) / denom) : 1.0;
-        const auto facL = static_cast<DType1>(1.0) - facR;
-        const auto yout = facL * yinp[j] + facR * yinp[j + 1];
-        return yout;
-    }
-
-    amrex::Abort("idx undefined");
-    exit(1);
+    static constexpr DType1 eps = 1.0e-8;
+    int j = idx.idx;
```
Comment (marchdf) [inline amr-wind/utilities/linear_interpolation.H:98]: `const int j`?

File: docs/sphinx/theory/theory.rst (hunk 10 above)
Comment (moprak-nrel) [reply docs/sphinx/theory/theory.rst:290]: Yup, these were showing the hand unrolled calcs, but those are no longer being used. The tensor form is almost exactly reflected in the code.

**Review by marchdf — APPROVED** [review]

**Pull request merged**